compiler/passes/src/write_transforming/statement.rs

```diff
@@ -0,0 +1,179 @@
+        // to mutably borrow `self` again below.
+        if let Some(members) = self.array_members.get(&name.name).cloned() {
+            for (i, member) in members.iter().cloned().enumerate() {
+                // Create a definition for each array index.
+                let index = Literal::integer(
+                    IntegerType::U8,
+                    i.to_string(),
+                    Default::default(),
+                    self.state.node_builder.next_id(),
+                );
+                self.state.type_table.insert(index.id(), Type::Integer(IntegerType::U32));
+                let access = ArrayAccess {
+                    array: name.into(),
+                    index: index.into(),
+                    span: Default::default(),
+                    id: self.state.node_builder.next_id(),
+                };
+                self.state.type_table.insert(access.id(), self.state.type_table.get(&member.id()).unwrap().clone());
+                let def = DefinitionStatement {
+                    place: DefinitionPlace::Single(member),
+                    type_: Type::Err,
+                    value: access.into(),
+                    span: Default::default(),
+                    id: self.state.node_builder.next_id(),
+                };
+                accumulate.push(def.into());
+                // And recurse - maybe its members are also written to.
+                self.define_variable_members(member, accumulate);
+            }
+        } else if let Some(members) = self.struct_members.get(&name.name) {
+            for (&field_name, &member) in members.clone().iter() {
+                // Create a definition for each field.
+                let access = MemberAccess {
+                    inner: name.into(),
+                    name: Identifier::new(field_name, self.state.node_builder.next_id()),
+                    span: Default::default(),
+                    id: self.state.node_builder.next_id(),
+                };
+                self.state.type_table.insert(access.id(), self.state.type_table.get(&member.id()).unwrap().clone());
+                let def = DefinitionStatement {
+                    place: DefinitionPlace::Single(member),
+                    type_: Type::Err,
+                    value: access.into(),
+                    span: Default::default(),
+                    id: self.state.node_builder.next_id(),
+                };
+                accumulate.push(def.into());
+                // And recurse - maybe its members are also written to.
+                self.define_variable_members(member, accumulate);
+            }
+        }
+    }
+}
+
+impl StatementReconstructor for WriteTransformingVisitor<'_> {
+    /// This is the only function where we do anything interesting - the rest just traverse,
+    /// combining statements.
+    fn reconstruct_assign(&mut self, input: AssignStatement) -> (Statement, Self::AdditionalOutput) {
```

Comment:
Is the interesting part the `reconstruct_assign_place`? Wasn't too clear from the comment.